From 0aa8138e0b08fb4e56e8edbe625b0997017a2ed6 Mon Sep 17 00:00:00 2001
From: Tamas K Lengyel <lengyelt@ainfosec.com>
Date: Tue, 19 Sep 2017 22:52:32 -0400
Subject: [PATCH 1/6] Add KEEP_DISCARDABLE_RELOC build option

Some efi applications expect the .reloc section to be present even after going
through the shim (ie. Xen). This option allows building the shim such that it
keeps it even if it was marked discardable.

Signed-off-by: Tamas K Lengyel <lengyelt@ainfosec.com>
---
 BUILDING      | 4 ++++
 Make.defaults | 4 ++++
 shim.c        | 6 +++++-
 3 files changed, 13 insertions(+), 1 deletion(-)

diff --git a/BUILDING b/BUILDING
index fb27821..c1e0ffc 100644
--- a/BUILDING
+++ b/BUILDING
@@ -59,5 +59,9 @@ Variables you could set to customize the build:
 - OSLABEL
   This is the label that will be put in BOOT$(EFI_ARCH).CSV for your OS.
   By default this is the same value as EFIDIR .
+- KEEP_DISCARDABLE_RELOC
+  This allows you to decide whether a .reloc section that was marked
+  discardable gets discarded or not. Some efi applications expect it to be
+  present after the shim (ie. Xen). By default it gets discarded.
 
 # vim:filetype=mail:tw=74
diff --git a/Make.defaults b/Make.defaults
index e11ab5a..ba01917 100644
--- a/Make.defaults
+++ b/Make.defaults
@@ -63,6 +63,10 @@ ifneq ($(origin REQUIRE_TPM), undefined)
 	CFLAGS  += -DREQUIRE_TPM
 endif
 
+ifneq ($(origin KEEP_DISCARDABLE_RELOC), undefined)
+    CFLAGS  += -DKEEP_DISCARDABLE_RELOC
+endif
+
 ifeq ($(ARCH),x86_64)
 	CFLAGS	+= -mno-mmx -mno-sse -mno-red-zone -nostdinc \
 		   $(CLANG_BUGS) -m64 \
diff --git a/shim.c b/shim.c
index 0015534..0a4c201 100644
--- a/shim.c
+++ b/shim.c
@@ -1377,7 +1377,11 @@ static EFI_STATUS handle_image (void *data, unsigned int datasize,
 			}
 		}
 
-		if (Section->Characteristics & EFI_IMAGE_SCN_MEM_DISCARDABLE) {
+		if ( (Section->Characteristics & EFI_IMAGE_SCN_MEM_DISCARDABLE)
+#if defined(KEEP_DISCARDABLE_RELOC)
+		    && RelocSection != Section
+#endif
+		) {
 			continue;
 		}
 
-- 
2.17.1

